curvefs/src/mds/topology/topology_manager.cpp

```diff
@@ -935,47 +935,169 @@ TopoStatusCode TopologyManager::GetCopysetMembers(
     CopySetInfo info;
-    auto copySetInfo = new curvefs::mds::topology::Copyset();
     if (topology_->GetCopySet(key, &info)) {
-        copySetInfo->set_poolid(info.GetPoolId());
-        copySetInfo->set_copysetid(info.GetId());
-        MetaServer ms;
-        for (auto msId : info.GetCopySetMembers()) {
+        auto valueCopysetInfo = new curvefs::mds::heartbeat::CopySetInfo();
+        valueCopysetInfo->set_poolid(info.GetPoolId());
+        valueCopysetInfo->set_copysetid(info.GetId());
+        // set peers
+        for (auto const& msId : info.GetCopySetMembers()) {
+            MetaServer ms;
             if (topology_->GetMetaServer(msId, &ms)) {
-                common::Peer* peer = copySetInfo->add_peers();
+                common::Peer* peer = valueCopysetInfo->add_peers();
                 peer->set_id(ms.GetId());
                 peer->set_address(BuildPeerIdWithIpPort(ms.GetInternalHostIp(),
                                                         ms.GetInternalPort()));
-
-                response->set_statuscode(TopoStatusCode::TOPO_OK);
             } else {
-                LOG(ERROR) << "GetMetaServer failed, metaserverId = " << msId;
-                response->set_statuscode(
+                LOG(ERROR) << "perrs: poolId=" << poolId
+                           << " copysetid=" << copysetId
+                           << " has metaServer error, metaserverId = " << msId;
+                copysetValue->set_statuscode(
                     TopoStatusCode::TOPO_METASERVER_NOT_FOUND);
             }
         }
+        valueCopysetInfo->set_epoch(info.GetEpoch());
+
+        // set leader peer
+        auto msId = info.GetLeader();
+        MetaServer ms;
+        auto peer = new common::Peer();
+        if (topology_->GetMetaServer(msId, &ms)) {
+            peer->set_id(ms.GetId());
+            peer->set_address(BuildPeerIdWithIpPort(ms.GetInternalHostIp(),
+                                                    ms.GetInternalPort()));
+        } else {
+            LOG(WARNING) << "leaderpeer: poolId=" << poolId
+                         << " copysetid=" << copysetId
+                         << " has metaServer error, metaserverId = " << msId;
+            copysetValue->set_statuscode(
+                TopoStatusCode::TOPO_METASERVER_NOT_FOUND);
+        }
+        valueCopysetInfo->set_allocated_leaderpeer(peer);
+
+        // set partitioninfolist
+        for (auto const& i : info.GetPartitionIds()) {
+            Partition tmp;
+            if (!topology_->GetPartition(i, &tmp)) {
+                LOG(WARNING)
+                    << "poolId=" << poolId << " copysetid=" << copysetId
+                    << " has pattition error, partitionId=" << i;
+                copysetValue->set_statuscode(
+                    TopoStatusCode::TOPO_PARTITION_NOT_FOUND);
+            } else {
+                auto partition = valueCopysetInfo->add_partitioninfolist();
+                partition->set_fsid(tmp.GetFsId());
```

Comment:
You can use proto2 interface to copy `Partition` directly.

Reply:
> 

fix